kafka: require prometheus for monitoring features

zmalik opened the pull request:
service monitor and other features of KUDO Kafka require prometheus to be installed

shaneutt commented on addons/kafka/1.x/kafka-2.yaml:
```diff
@@ -15,6 +15,8 @@ spec:
   requires:
     - matchLabels:
         kubeaddons.mesosphere.io/name: zookeeper
+    - matchLabels:
```
Since this significantly limits how one can use and test Kafka, should we add a comment here explaining the "why" to any future readers?

zmalik replied:
yeah, let me add
KUDO Kafka monitoring relies on the `prometheus` operator and uses objects like service monitor.
In case this is an `enterprise` addon with possible support. We wouldn't recommend or provide a way for users to install KUDO Kafka without monitoring which means without `prometheus` installed. As this would limit considerably any effort to effectively monitor KUDO Kafka clusters.

shaneutt replied:
:+1:  check this comment out for spark too, for possible relevance:
https://github.com/mesosphere/kubeaddons-enterprise/pull/18#discussion_r374128574

shaneutt commented:
@zmalik this wont pass CI because the dependency on prometheus has not been met in the CI setup.

Normally you would want to add `prometheus` to the same [testing group](https://github.com/mesosphere/kubeaddons-enterprise/blob/8c74c59c222484b13d501960f944ceacaf369a72/test/groups.yaml) as Kafka, but that also isn't set up to work right out of the box right now, because we're not configured to deploy external addons in test groups.

I think the next step should be to support external addons in test groups (e.g. an http URL as valid in the groups) and point to `kubernetes-base-addons` master prometheus. This should be a matter of adding the functionality to `Create` from URLs in the testgroup CI code.

Let us know if this sounds good to you, if you have any questions, and if you need any help!

zmalik commented:
@shaneutt let me take a stab and I will update this PR

shaneutt commented:
> @shaneutt let me take a stab and I will update this PR

Sounds good, for your consideration: it may be viable/reasonable to simply hook CI to remove the dependency and manually inject the CRD, we're discussing that kind of a solution in another PR.

zmalik commented:
so installing prometheus chart is failing due to https://jira.d2iq.com/browse/D2IQ-63819 
and installing CRD for `servicemonitor` alone isn't enough as kubeaddons wait for prometheus addon to be installed and be healthy if we add it as a requirement.

zmalik commented:
@shaneutt I added support for remote addons living in https://github.com/mesosphere/kubernetes-base-addons, on a retry basis. In the end to not use it as, prometheus chart brings too much noise, even if I disable everything except the operator. 
And also moved the override logic to `groups.yaml`
And finally adding an option to remove dependency from the addon just for testing reasons using `group.yaml` to avoid too much custom logic in go code

shaneutt commented on test/addons_test.go:
```diff
@@ -19,17 +17,24 @@ import (
+
+type AddonTestConfiguration struct {
+	Name               string   `json:"name,omitempty" yaml:"name,omitempty"`
+	Override           string   `json:"override,omitempty" yaml:"override,omitempty"`
```
:+1: in fact let's make sure we add godoc comments for all public symbols.

shaneutt commented on test/addons_test.go:
```diff
@@ -19,17 +17,24 @@ import (
+type AddonTestConfiguration struct {
+	Name               string   `json:"name,omitempty" yaml:"name,omitempty"`
+	Override           string   `json:"override,omitempty" yaml:"override,omitempty"`
+	RemoveDependencies []string `json:"removeDependencies,omitempty" yaml:"removeDependencies,omitempty"`
```
++

zmalik replied:
done!

shaneutt commented on test/addons_test.go:
```diff
@@ -171,34 +176,37 @@ func testgroup(t *testing.T, groupname string) error {
+		if err != nil {
+			fmt.Printf("Cannot find the addon [%s] in local repository\n", addonConfig.Name)
+			fmt.Printf("Will try to fetch [%s] from remote repository addon\n", addonConfig.Name)
+			remoteRepo, err := git.NewRemoteRepository(constants.KubeAddonsConfigsRepository, "master", "origin")
```
I agree with Jan, let's fail fast and keep small defensive dimensions to our test code, at least for this iteration until we've had more operational experience with it in CI.

zmalik replied:
Im removing it altogether as nobody is using it. I know this code works and we can re-add it by enabling in groups.yaml

shaneutt commented on test/groups.yaml:
```diff
@@ -16,19 +16,35 @@
+- name: "kafka"
+  # remove promethues dependency for CI
+  # https://jira.d2iq.com/browse/D2IQ-63819
+  removeDependencies:
```
I do like how this improves the composition of the testing groups. :+1:

zmalik commented:
@shaneutt @nfnt can you please re-approve, I had to sign commits to and push again
